### PR TITLE
Add initial notice

### PR DESCRIPTION
**Pull request by felixheidecke**
bildschirmfoto-2017-11-09-um-18 55 52

Fixes: https://github.com/owncloud/market/issues/179

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -7,15 +7,45 @@
 
 			main.uk-width-expand
 				router-view
+
+		#scaleout-notice-modal(uk-modal="bg-close : false")
+			.uk-modal-dialog.uk-margin-auto-vertical
+				button.uk-modal-close-default(type='button', uk-close)
+				.uk-modal-header
+						h2.uk-modal-title Important notice
```
Comment (DeepDiver1975) [inline src/App.vue]: translation support is missing

Comment (PVince81) [reply src/App.vue]: indeed, and translation support for handlebars is still missing too

Comment (felixheidecke) [reply src/App.vue]: Will be there shortly ;-)

File: src/App.vue [hunk 2]
```diff
@@ -7,15 +7,45 @@
 
 			main.uk-width-expand
 				router-view
+
+		#scaleout-notice-modal(uk-modal="bg-close : false")
```
Comment (hurradieweltgehtunter) [inline src/App.vue]: Please do not use modals. Blank screen on initial market app startup where the user is asked to activate or disable the market app or similar. But no modals.

**Comment by DeepDiver1975** [comment]
no idea if this is still needed.

https://github.com/owncloud/core/pull/29458 will introduce an explicit config flag which allows us to find out if owncloud is clustered or not.

Please let me know if you need some api calls (market frontend and market backend) to query for this setting. I could think of adding notification on top of the page which tell the user that installation/uprade is not possible.

installing and upgrading will be disallowed by core

**Comment by DeepDiver1975** [comment]
... and disable the install and update buttons

**Comment by PVince81** [comment]
https://github.com/owncloud/core/pull/29458 will be in RC, please adjust this PR accordingly.

You can still display whatever message, but only if said setting is set.

@DeepDiver1975 we could expose the setting over capabilities API ?

**Comment by PVince81** [comment]
Capabilities API results are available by default in the JS array returned by `OC.getCapabilities()`, synchronously. (loaded at page load)

**Comment by hurradieweltgehtunter** [comment]
@felixheidecke could you adapt to @DeepDiver1975 's comment for cluster flag?
ty

**Comment by felixheidecke** [comment]
> @felixheidecke could you adapt to @DeepDiver1975 's comment for cluster flag?

@hurradieweltgehtunter Sure I can, but it will stray from the original idea. The message would pop-up every time you open the market app and say something like:

_"Since this is a clustered setup, installing and updating apps will not work. Have fun anyway."_

If that's okay, I'll need an API route telling me about the setup :-)

**Comment by PVince81** [comment]
@felixheidecke instead of an API route what we could do is to put said code into a separate JS file and only load said file from PHP (include it) whenever the mode is set. But I suspect that this approach is not compatible with webpack / vue JS ?

**Comment by felixheidecke** [comment]
@PVince81 sorry. That won't work in this case :-|

File: src/App.vue (hunk 1 above)
Comment (felixheidecke) [reply src/App.vue]: Translation support exists but has not been used in this early state. Will wait for the final text.

**Comment by hurradieweltgehtunter** [comment]
Isn't there a global settings variable or sth else, which is delivered to the frontend and where the cluster flag could be included?
@felixheidecke read this flag and dependent on this enable/disable the whole market app (also in BE) ?

**Comment by PVince81** [comment]
> global settings variable 

either we need to add it to the capabilities API, then it will be available directly.

else we need to modify the PHP template to inject it through the DOM in a "data-xxx" attribute

**Comment by PVince81** [comment]
> else we need to modify the PHP template to inject it through the DOM in a "data-xxx" attribute

would this be acceptable ? else we'll need another core PR to add the flag to the capabilities

**Comment by felixheidecke** [comment]
@PVince81 Would this happen somewhere in [the template](https://github.com/owncloud/market/blob/master/templates/index.php)??

**Comment by PVince81** [comment]
oops, so the template has no DOM elements at all ?

**Comment by felixheidecke** [comment]
Eeeeexactly :-)

**Comment by felixheidecke** [comment]
Feels hacky to use HTML as a vehicle to transport infos from PHP to JS :-/

**Comment by PVince81** [comment]
> Feels hacky to use HTML as a vehicle to transport infos from PHP to JS :-/

don't look into core PHP code then...

Ok, looks like the capabilities is the only way. Adding a new endpoint feels like overkill.

**Comment by PVince81** [comment]
I've added/modifed an endpoint as discused with @felixheidecke, see https://github.com/owncloud/market/pull/184/commits/203a1f66170c68ab22116811254e1c7e2649f3e5

**Comment by felixheidecke** [comment]
@hurradieweltgehtunter @pmaier1 can I get a propper text to display?

**Comment by pmaier1** [comment]
> @hurradieweltgehtunter @pmaier1 can I get a propper text to display?

Yes, I could make something up, but wait a sec.

1. To my knowledge the market app in it's current state should _never_ be used in a clustered setup.
=> Why would we then at all provide the ability to use it when operating in a clustered setup? In other words: With the suggested approach we say "Don't ever use this, here is information why and so on and so on - but - here it is, have fun :+1:"
2. Now that we 'know' when an instance is in cluster mode I would rather vote for
- Disabling the market app when in cluster mode or
- Leaving it enabled and instead of opening it, just show a hint: "The Market app is currently unsupported in clustered setups. More information here"

Wouldn't this be better for admins and a lot easier for us? What am I missing?

**Comment by DeepDiver1975** [comment]
Even in Cluster Mode the Admin can have a look at available Apps or receive notifications.
The Market App is not only about installing Apps.

**Comment by hurradieweltgehtunter** [comment]
Maybe disable install/update buttons? Or show a hint "You're in a clustered setup. Installing/Updating apps goes like this:
1. blablabla
2. blablabla"

And don't forget the CLI!

**Comment by pmaier1** [comment]
> Even in Cluster Mode the Admin can have a look at available Apps or receive notifications.
The Market App is not only about installing Apps.

Ok, good point. Update notifications are very important, of course. Installing/updating should be prevented in any case IMO.

Text proposal:

```
You're running ownCloud in a clustered setup. Installing or upgrading apps via the Market app is currently not supported. Please follow the [manual process](https://doc.owncloud.com/server/10.0/admin_manual/upgrading/marketplace_apps.html#clustered-multi-server-environment).
You can nevertheless browse the available apps and will receive update notifications.
```

I would add a sentence "For details and further instructions, please see.." but I could not find anything further in the docs. Not sure if this is needed anyway. Actually the text and the doc link give all information needed.

**Comment by felixheidecke** [comment]
The config route returns `canInstall` either true or false. Are we sure that canInstall is false **only** if its a cluster setup? Will it also be false if the apps folder is not writable? Or maybe a few other reasons? @PVince81 

I'm asking since this is my trigger to display the cluster warning.

**Comment by PVince81** [comment]
@felixheidecke good point. Looking at the core code for AppManager::canInstall, I see that i also checks writeability of the apps folder. So might need to tweak the message to say "either you're in a clustered env or apps folder is not writeable".

**Comment by felixheidecke** [comment]
Could we supply a message/string alongside that says something like 'cluster setup' or 'apps folder not writable' or whatever?

**Comment by DeepDiver1975** [comment]
Can we please keep it simple? 
Installation/updating is not allowed. Game over.

**Comment by DeepDiver1975** [comment]
Please use the marketservice::caninstall

**Comment by PVince81** [comment]
> Please use the marketservice::caninstall

Done

**Comment by felixheidecke** [comment]
> Installation/updating is not allowed. Game over.

Is this the new message to be displayed? 😁  @pmaier1 @DeepDiver1975

**Comment by PVince81** [comment]
Maybe don't display a warning at all, but grey out the install buttons.

And close the install button, display a message like "App cannot be installed over the web UI either because this is a clustered setup or the web server has no permissions to write to the apps folder".

Now if we consider the separation between OC admin and sysadmin, the OC admin might not need to know about apps folder permissions and a message like "App cannot be installed over the web UI, please check with the system administrator" + link to docs ?

**Comment by PVince81** [comment]
or => "App cannot be installed over the web UI *in this setup*"

**Comment by felixheidecke** [comment]
If the API is smart enough, it should already send a `canInstall: false` per app.

**Comment by DeepDiver1975** [comment]
> If the API is smart enough, it should already send a canInstall: false per app.

what would be possible

**Comment by DeepDiver1975** [comment]
any travis hates us - please fix - maybe related to #186  ?????

**Comment by felixheidecke** [comment]
I'll go ahead and show a warning which can be dismissed.

**Comment by felixheidecke** [comment]
See https://github.com/owncloud/market/pull/184#issue-270644841 for updated notice.

**Comment by PVince81** [comment]
@felixheidecke some trouble with travis / node ?

**Comment by PVince81** [comment]
Assuming you were happy with the backend code @DeepDiver1975 ?

Who reviews the frontend code ?

**Comment by felixheidecke** [comment]
> Who reviews the frontend code?

That's always a goooood question. @hurradieweltgehtunter and @tboerger know vue. 🤔

**Comment by PVince81** [comment]
@felixheidecke please retest this properly, not only the new switch but also the license related code.

The server backend has changed and there's a key called "licenseKeyAvailable" but I don't see it read anywhere in the UI code. See https://github.com/owncloud/market/pull/184/files#diff-6248f15975b779eff847acf5c338dd48R277

**Pull request merged**